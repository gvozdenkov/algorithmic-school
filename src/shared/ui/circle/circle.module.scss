@use '#styles/media-bp-mixins.scss' as *;

.circle {
  position: relative;

  display: flex;
  flex-direction: column;
  align-items: center;

  box-sizing: border-box;
  width: fit-content;

  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell,
    'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;
  font-size: 18px;
  font-weight: normal;
  font-style: normal;
  line-height: 22px;
  color: var(--text-color-input);

  &__shape {
    display: flex;
    align-items: center;
    justify-content: center;

    box-sizing: border-box;
    aspect-ratio: 1;
    width: var(--circle-radius);

    border-radius: 50%;

    &_small {
      @extend .circle__shape;

      width: calc(var(--circle-radius) * 0.7);
    }

    &_state_default {
      @extend .circle__shape;

      border: var(--circle-border) solid var(--default-color);
    }

    &_state_changing {
      @extend .circle__shape;

      border: var(--circle-border) solid var(--changing-color);
    }

    &_state_modified {
      @extend .circle__shape;

      border: var(--circle-border) solid var(--modified-color);
    }
  }

  &__text {
    font-family: Montserrat, sans-serif;
    font-size: 18px;
    font-weight: 600;
    font-style: normal;
    line-height: 22px;
    text-transform: uppercase;
  }

  &__head {
    &_type_string {
      @extend .absolute;

      top: -30px;
    }

    &_type_element {
      @extend .absolute;

      top: -70px;
    }
  }

  &__tail {
    &_type_string {
      @extend .absolute;

      bottom: -60px;
    }

    &_type_element {
      @extend .absolute;

      bottom: -100px;
    }
  }

  &__index {
    @extend .absolute;

    bottom: -30px;
  }
}

.absolute {
  position: absolute;
  left: 50%;
  transform: translate(-50%);
}

@include mobile {
  .circle {
    &__text {
      font-size: 14px;
      font-weight: 400;
      line-height: 18px;
      text-transform: uppercase;
    }
  }
}
