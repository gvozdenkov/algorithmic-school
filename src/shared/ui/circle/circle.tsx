import React from 'react';
import s from './circle.module.scss';
import { ElementState } from '#shared/types';
import clsx from 'clsx';

type CircleProps = {
  state?: ElementState;
  letter?: string;
  head?: string | React.ReactElement | null | false;
  index?: number;
  tail?: string | React.ReactElement | null | false;
  tailType?: 'string' | 'element';
  extraClass?: string;
  isSmall?: boolean;
};

export const Circle = ({
  state = 'default',
  letter,
  head,
  index,
  tail,
  extraClass = '',
  isSmall,
}: CircleProps) => {
  return (
    <div className={clsx(s.circle, { [extraClass]: !!extraClass })}>
      <div
        className={clsx(
          'text mb-4',
          {
            [s.circle__head_type_string]: typeof head === 'string',
          },
          {
            [s.circle__head_type_element]: typeof head !== 'string',
          },
        )}>
        {head}
      </div>
      <div
        className={clsx(s[`circle__shape_state_${state}`], {
          [s.circle__shape_small]: isSmall,
        })}>
        <p className={clsx('text', s.circle__text)}>{letter}</p>
      </div>
      <p className={clsx('text mt-4', s.circle__index)}>{index?.toString()}</p>
      <div
        className={clsx(
          'text mt-4',
          { [s.circle__tail_type_string]: typeof tail === 'string' },
          { [s.circle__tail_type_element]: typeof tail !== 'string' },
        )}>
        {tail}
      </div>
    </div>
  );
};
