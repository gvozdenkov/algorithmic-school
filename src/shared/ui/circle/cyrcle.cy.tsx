import { colorSwitchToRGB } from '#shared/lib';

import { Circle } from './circle';

describe('circle', () => {
  it('should render empty circle in all states', () => {
    cy.mount(<Circle letter='a' />);
    cy.getBySel('circleShape').should('have.css', {
      borderColor: `${colorSwitchToRGB('default')}`,
    });

    // cy.mount(<Circle state='changing' />);
    // cy.getBySel('circleShape').should(
    //   'have.css',
    //   'border-color',
    //   `${colorSwitchToRGB('changing')}`,
    // );

    // cy.mount(<Circle state='modified' />);
    // cy.getBySel('circleShape').should(
    //   'have.css',
    //   'border-color',
    //   `${colorSwitchToRGB('modified')}`,
    // );
  });

  it('should render small circle', () => {
    cy.mount(<Circle isSmall />);
    cy.getBySel('circleShape').should('have.css', { width: '56px' });
  });

  it('should render circle with text, index, head and tail', () => {
    cy.mount(<Circle letter='a' index={1} head='head' tail='tail' />);
    cy.contains('a');
    cy.contains('1');
    cy.contains('head');
    cy.contains('tail');
  });
});
