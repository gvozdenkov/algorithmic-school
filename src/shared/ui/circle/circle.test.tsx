import { create } from 'react-test-renderer';

import { Circle } from './circle';

describe('Circle ui', () => {
  it('without text', () => {
    const tree = create(<Circle />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('with text', () => {
    const tree = create(<Circle letter='a' />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('with head', () => {
    const tree = create(<Circle head={'head'} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('with element in head', () => {
    const tree = create(<Circle head={<Circle letter='a' />} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('with tail', () => {
    const tree = create(<Circle tail='tail' />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('with element in tail', () => {
    const tree = create(<Circle tail={<Circle letter='a' />} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('with index', () => {
    const tree = create(<Circle index={1} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('with isSmall prop', () => {
    const tree = create(<Circle isSmall />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('state: default', () => {
    const tree = create(<Circle state='default' />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('state: changing', () => {
    const tree = create(<Circle state='changing' />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('state: modified', () => {
    const tree = create(<Circle state='modified' />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
