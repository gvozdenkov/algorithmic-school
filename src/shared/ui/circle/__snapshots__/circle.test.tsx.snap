// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle ui state: changing 1`] = `
<div
  className="circle"
>
  <div
    className="text mb-4 circle__head_type_element"
  />
  <div
    className="circle__shape_state_changing"
    data-test="circleShape"
  >
    <p
      className="text circle__text"
      data-test="circleText"
    />
  </div>
  <p
    className="text mt-4 circle__index"
  />
  <div
    className="text mt-4 circle__tail_type_element"
  />
</div>
`;

exports[`Circle ui state: default 1`] = `
<div
  className="circle"
>
  <div
    className="text mb-4 circle__head_type_element"
  />
  <div
    className="circle__shape_state_default"
    data-test="circleShape"
  >
    <p
      className="text circle__text"
      data-test="circleText"
    />
  </div>
  <p
    className="text mt-4 circle__index"
  />
  <div
    className="text mt-4 circle__tail_type_element"
  />
</div>
`;

exports[`Circle ui state: modified 1`] = `
<div
  className="circle"
>
  <div
    className="text mb-4 circle__head_type_element"
  />
  <div
    className="circle__shape_state_modified"
    data-test="circleShape"
  >
    <p
      className="text circle__text"
      data-test="circleText"
    />
  </div>
  <p
    className="text mt-4 circle__index"
  />
  <div
    className="text mt-4 circle__tail_type_element"
  />
</div>
`;

exports[`Circle ui with element in head 1`] = `
<div
  className="circle"
>
  <div
    className="text mb-4 circle__head_type_element"
  >
    <div
      className="circle"
    >
      <div
        className="text mb-4 circle__head_type_element"
      />
      <div
        className="circle__shape_state_default"
        data-test="circleShape"
      >
        <p
          className="text circle__text"
          data-test="circleText"
        >
          a
        </p>
      </div>
      <p
        className="text mt-4 circle__index"
      />
      <div
        className="text mt-4 circle__tail_type_element"
      />
    </div>
  </div>
  <div
    className="circle__shape_state_default"
    data-test="circleShape"
  >
    <p
      className="text circle__text"
      data-test="circleText"
    />
  </div>
  <p
    className="text mt-4 circle__index"
  />
  <div
    className="text mt-4 circle__tail_type_element"
  />
</div>
`;

exports[`Circle ui with element in tail 1`] = `
<div
  className="circle"
>
  <div
    className="text mb-4 circle__head_type_element"
  />
  <div
    className="circle__shape_state_default"
    data-test="circleShape"
  >
    <p
      className="text circle__text"
      data-test="circleText"
    />
  </div>
  <p
    className="text mt-4 circle__index"
  />
  <div
    className="text mt-4 circle__tail_type_element"
  >
    <div
      className="circle"
    >
      <div
        className="text mb-4 circle__head_type_element"
      />
      <div
        className="circle__shape_state_default"
        data-test="circleShape"
      >
        <p
          className="text circle__text"
          data-test="circleText"
        >
          a
        </p>
      </div>
      <p
        className="text mt-4 circle__index"
      />
      <div
        className="text mt-4 circle__tail_type_element"
      />
    </div>
  </div>
</div>
`;

exports[`Circle ui with head 1`] = `
<div
  className="circle"
>
  <div
    className="text mb-4 circle__head_type_string"
  >
    head
  </div>
  <div
    className="circle__shape_state_default"
    data-test="circleShape"
  >
    <p
      className="text circle__text"
      data-test="circleText"
    />
  </div>
  <p
    className="text mt-4 circle__index"
  />
  <div
    className="text mt-4 circle__tail_type_element"
  />
</div>
`;

exports[`Circle ui with index 1`] = `
<div
  className="circle"
>
  <div
    className="text mb-4 circle__head_type_element"
  />
  <div
    className="circle__shape_state_default"
    data-test="circleShape"
  >
    <p
      className="text circle__text"
      data-test="circleText"
    />
  </div>
  <p
    className="text mt-4 circle__index"
  >
    1
  </p>
  <div
    className="text mt-4 circle__tail_type_element"
  />
</div>
`;

exports[`Circle ui with isSmall prop 1`] = `
<div
  className="circle"
>
  <div
    className="text mb-4 circle__head_type_element"
  />
  <div
    className="circle__shape_state_default circle__shape_small"
    data-test="circleShape"
  >
    <p
      className="text circle__text"
      data-test="circleText"
    />
  </div>
  <p
    className="text mt-4 circle__index"
  />
  <div
    className="text mt-4 circle__tail_type_element"
  />
</div>
`;

exports[`Circle ui with tail 1`] = `
<div
  className="circle"
>
  <div
    className="text mb-4 circle__head_type_element"
  />
  <div
    className="circle__shape_state_default"
    data-test="circleShape"
  >
    <p
      className="text circle__text"
      data-test="circleText"
    />
  </div>
  <p
    className="text mt-4 circle__index"
  />
  <div
    className="text mt-4 circle__tail_type_string"
  >
    tail
  </div>
</div>
`;

exports[`Circle ui with text 1`] = `
<div
  className="circle"
>
  <div
    className="text mb-4 circle__head_type_element"
  />
  <div
    className="circle__shape_state_default"
    data-test="circleShape"
  >
    <p
      className="text circle__text"
      data-test="circleText"
    >
      a
    </p>
  </div>
  <p
    className="text mt-4 circle__index"
  />
  <div
    className="text mt-4 circle__tail_type_element"
  />
</div>
`;

exports[`Circle ui without text 1`] = `
<div
  className="circle"
>
  <div
    className="text mb-4 circle__head_type_element"
  />
  <div
    className="circle__shape_state_default"
    data-test="circleShape"
  >
    <p
      className="text circle__text"
      data-test="circleText"
    />
  </div>
  <p
    className="text mt-4 circle__index"
  />
  <div
    className="text mt-4 circle__tail_type_element"
  />
</div>
`;
